### ISSUE TITLE
Add Default UUID on External Project Nodes

### ISSUE THREAD

**Issue by ClxS** [premake/premake-core#1330]
**What seems to be the problem?**
It seems like there aren't any specific requirements when generating an external project UUID, so having to manually make one for each project feels like busy work when doing things such as including a large number of projects which premake doesn't currently support (.NET Core)

**What might be a solution?**
Project's could have their UUID initialised to uuid = os.uuid(name) in the external(..) func.

**Comment by starkos** [comment]
I was confused by this on first read: I think the intention here is to set a default UUID for non-Visual Studio projects. A better solution, perhaps, would be to move the UUID check in to the VS exporter, which is the only place that value is actually required (I believe).

**Comment by samsinsane** [comment]
@starkos I don't think it's _required_ in the VS exporter either though? To my understanding the UUID is only used in the solution file, so it doesn't matter what it's set to, as long as it isn't the same as another project.

**Comment by ClxS** [comment]
@samsinsane That's right, sorry I might have worded the initial message ambiguously. My issue was mainly when generating Visual Studio solutions. As the only requirements for the UUID in the solution are:
1) It's unique
2) It's deterministic so VS doesn't reload every project on regenerate
Couldn't the UUID be formed using the name of the project?

**Comment by starkos** [comment]
Hmm…I never realized that was the case, I always thought they had to match up with the identifier in the project file. If that's true, you can get a deterministic UUID with `os.uuid("path/to/project")` (or whatever project ID string you want to use).

**Comment by samsinsane** [comment]
> I always thought they had to match up with the identifier in the project file

Oh I see, I guess my experience is only with non-C++ projects but looking at the `.vcproj` and `.vcxproj` generators, having Premake generate the UUID could be bad. `links { "<external project>" }` seems like it would generate a bad reference, the external project would have a UUID of X internally but we'd be referencing Y - [`here`](https://github.com/premake/premake-core/blob/49175231feefd4a74e0aa66d7235147584480b20/modules/vstudio/vs200x_vcproj.lua#L1386-L1408) for `.vcproj` and [`here`](https://github.com/premake/premake-core/blob/49175231feefd4a74e0aa66d7235147584480b20/modules/vstudio/vs2010_vcxproj.lua#L2401-L2403) for `.vcxproj`. I guess there's a possibility that the `.vcproj` and `.vcxproj` reference the UUID specified in the `.sln` when linking, but we'd need to validate this for all versions of VS since VS2005.

Perhaps we could add a magic value that will generate a deterministic UUID for when the user knows it won't matter for linking, or the project doesn't internally contain a UUID. `uuid "auto"` perhaps? This way the user is always aware of the UUID requirement but has the option of saying "I don't care", thoughts?

**Comment by starkos** [comment]
I'm okay with the idea of `uuid "auto"`, but off the top of my head I don't see how you would make it emit a deterministic value?

**Comment by samsinsane** [comment]
I was just assuming we'd use `os.uuid("path/to/project")` or something, as you suggested above?

**Comment by starkos** [comment]
Right, but if the user scripts this:

```lua
project "MyProject"
    uuid "auto"
```

Where does "auto" get converted into a real UUID? I guess the `uuid()` function could peek into `api.scope` to fetch the project name, though that feels a little "dirty". Or it could be assigned during baking but that breaks up the UUID logic, which would ideally all be in one place?

**Comment by starkos** [comment]
Another idea: maybe we should remove the UUID check for external projects, and let the individual actions do it when it is required (i.e. in the Visual Studio exporter)? That way you could simply not specify it if you don't need it.

**Comment by ratzlaff** [comment]
(Just throwing out ideas here)
Is it worth trying to parse the referenced project file to find its UUID?

**Comment by samsinsane** [comment]
> Another idea: maybe we should remove the UUID check for external projects, and let the individual actions do it when it is required (i.e. in the Visual Studio exporter)? That way you could simply not specify it if you don't need it.

I like that idea :+1:

> Is it worth trying to parse the referenced project file to find its UUID?

I had thought of this and I genuinely like the idea, but I feel the complexity would be too much - I generally use the external project system for unsupported project types. Parsing for any UUID wouldn't work if they reference other projects, or if they don't have a UUID at all. I guess a combination of parsing and generating one could be good?

**Comment by starkos** [comment]
In theory we could. There is always the chance that the project doesn't exist at the time Premake is being run though. I think I'd prefer to let the user specify it, personally.
